tensorflow/core/framework/memory_types.cc

```diff
@@ -104,37 +108,58 @@ Status MemoryTypesForNode(const OpRegistryInterface* op_registry,
     return it != ndef.attr().end() && it->second.b();
   }();
 
-  bool has_kernel_def = status.ok() && !IsFunctionCallOp(ndef.op());
+  bool is_fn = IsFunctionCallOp(ndef.op());
+  bool has_kernel_def = status.ok() && !is_fn;
   auto host_memory_required = [&](const DataType& dt) {
     bool int32_on_device =
         has_kernel_def || device_type.type_string() == "TPU" || has_xla_compile;
     return DataTypeAlwaysOnHost(dt) || (dt == DT_INT32 && !int32_on_device);
   };
 
-  if (has_kernel_def) {
+  //  Edge cases:
+  //  1. If[Tcond=DT_BOOL, Tin=[DT_FLOAT,DT_INT32], Tout=[DT_FLOAT,DT_INT32]]
+  //     * Tcond marked HostMemory by kernel_def
+  //     * Tin[1] should be on host since it's INT32
+  //  2. ResourceGather[Tindices=DT_INT32](resource, indices)
+  //     * 'resource' marked HostMemory on the host
+  //     *  'indices' must be on the device (because the kernel expects it so)
+  //       - therefore, ignore the type
+
+  // Set all datatype to DEVICE_MEMORY by default, later on change it to
+  // HOST_MEMORY where it is required by the datatype.
+
+  if (has_kernel_def || is_fn) {
     // Gets the input/output names and their corresponding endpoint ranges.
     NameRangeMap inp_names;
     NameRangeMap out_names;
     TF_RETURN_IF_ERROR(
         NameRangesForNode(ndef, *op_def, &inp_names, &out_names));
 
     // Now that we know the size, fill with the default 'DEVICE_MEMORY'.
-    inp_mtypes->resize(GetTotal(inp_names), DEVICE_MEMORY);
-    out_mtypes->resize(GetTotal(out_names), DEVICE_MEMORY);
-
-    // Fills in host memory types based on the kernel def.
-    const auto& from_proto = kdef->host_memory_arg();
-    std::vector<string> host_memory_args(from_proto.begin(), from_proto.end());
-    MemoryTypesHelper(inp_names, &host_memory_args, inp_mtypes);
-    MemoryTypesHelper(out_names, &host_memory_args, out_mtypes);
-    if (!host_memory_args.empty()) {
-      return errors::InvalidArgument(
-          "HostMemory args '", absl::StrJoin(host_memory_args, "', '"),
-          "' not found in OpDef: ", SummarizeOpDef(*op_def));
+    // Stick to old behavior (mark named edges only) except for function calls
+    // (since some kernels incorrectly rely on INT32 being on the GPU)
+    if (is_fn) {
+      inp_mtypes->resize(inp_dtypes.size(), DEVICE_MEMORY);
+      out_mtypes->resize(out_dtypes.size(), DEVICE_MEMORY);
+    } else {
+      inp_mtypes->resize(GetTotal(inp_names), DEVICE_MEMORY);
+      out_mtypes->resize(GetTotal(out_names), DEVICE_MEMORY);
+    }
+
+    // Fills in host memory types based on the kernel def
+    if (kdef != nullptr) {  // can this ever be false?
```

Comment:
"can this ever be false?" -> Not in this branch of the if, but kdef will be false for functions that are called by setting the op type to the function name. We mostly use that for calling from eager, but there are graphs/unit tests floating around that use it in graphs too. I think it's safe to assume everything else has a kernel and remove this check.

Higher-level question: given that If/While/PartitionedCall/etc. do have kernels, should we be respecting the kernel annotations rather than special-casing int32 for them in the host_memory_required lambda? Did we just not properly annotate the int32 kernels yet?

Reply:
We need to do both.
Some inputs of If/While/etc. go direct into the corresponding kernel, and we mark them as host by looking at the annotations.
Other are passed through to subgraphs untouched. If I remember correctly (and again, it's been 6 months), what used to happen was that some DT_INT32 tensor was passed to If or While, and then materialized inside a subgraph as an output of an Arg(?) op that connected to a Less op. Less op expected host memory input, but If op was the one that determined tensor's location, and, as originally implemented, it did not mark the tensor as host.

Reply:
So what we really want to do is inspect the function and find the kernel(s) that eventually consume the tensor, check if they have a host memory annotation, and follow that. Or possibly easier, leave the tensor where it is an lazily copy it if necessary. Assuming that kernels will want int32-dtype tensors on host is pretty sad.

@qqfish we have lazy tensor copy for function calls from eager (IIUC). Do you know if it would be feasible for us to do something similar for tensors passed from graph mode into a functional op's kernel?

Reply:
I am not sure if the information about the ultimate consumer is even available at the time this is executed (the consumer may be in a different graph). 
Keeping the tensor in its default location  is suboptimal, because most consumers of DT_INT32 expect it to be on host (because it is a shape or a single integer.)
The optimal and robust solution would be to put INT32 outputs of functional ops on host, and to add checking when subgraphs are connected, in case it is actually expected on the device.

Reply:
Not "copy it to the device", but "don't do any copies until we hit a real kernel". Functional operations shouldn't have an opinion about where their inputs are, since kernels are free to consume int32 tensors on-device. And likewise they should leave their outputs where they are.

I assume we get this behavior when we inline control flow operations (which happens almost always for PartitionedCall/While/If). It'd be sad if failure to inline lead to special-casing int32. Just want to assess feasibility.

Reply:
Right now, lazy tensor copy in eager only works for cross-host tensor copy. And I don't think we should reuse here because I think we should only use lazy tensor copy in eager mode since in eager mode tensors are copied implicitly for usability. In graph mode, it makes more sense to let the placer to insert Send/Recv ops explicitly since it can make graph debugging easier.

So this change actually make sense to me. If we are going to support more functional op, I think we should consider adding a new kernel/op registration flag instead of hard coding op name here.

Reply:
The placer could still insert send/recv ops explicitly, it's just a question of whether we insert them before a functional op or in the body of the functional op's function (or both). So probably not the same infrastructure as eager lazy tensor copy, but instead just some bookkeeping to let functional ops know about host/device and let them forward that to their calls to the placer for the functions they're wrapping.

We allow kernels to produce/consume int32 tensors on device, so it'd be nice if feeding an int32 device tensor through a functional op didn't change that.